libraries/ESP8266WebServer/examples/FSBrowser/FSBrowser.ino

```diff
@@ -41,197 +67,534 @@ FS* filesystem = &LittleFS;
+    // Always return names without leading "/"
+    if (dir.fileName()[0] == '/') {
+      output += &(dir.fileName()[1]);
+    } else {
+      output += dir.fileName();
+    }
+
+    output += "\"}";
   }
+
+  // send last string
+  output += "]";
+  server.sendContent(output);
+  server.chunkedResponseFinalize();
 }
 
-void handleFileDelete() {
-  if (server.args() == 0) {
-    return server.send(500, "text/plain", "BAD ARGS");
+
+/*
+   Read the given file from the filesystem and stream it back to the client
+*/
+bool handleFileRead(String path) {
+  DBG_OUTPUT_PORT.println(String("handleFileRead: ") + path);
+  if (!fsOK) {
+    returnFail("FS INIT ERROR");
+    return true;
   }
-  String path = server.arg(0);
-  DBG_OUTPUT_PORT.println("handleFileDelete: " + path);
-  if (path == "/") {
-    return server.send(500, "text/plain", "BAD PATH");
+
+  if (path.endsWith("/")) {
+    path += "index.htm";
   }
-  if (!filesystem->exists(path)) {
-    return server.send(404, "text/plain", "FileNotFound");
+
+  String contentType;
+  if (server.hasArg("download")) {
+    contentType = "application/octet-stream";
+  } else {
+    contentType = getContentType(path);
   }
-  filesystem->remove(path);
-  server.send(200, "text/plain", "");
-  path.clear();
+
+  if (!fileSystem->exists(path)) {
+    // File not found, try gzip version
+    path = path + ".gz";
+  }
+  if (fileSystem->exists(path)) {
+    File file = fileSystem->open(path, "r");
+    if (server.streamFile(file, contentType) != file.size()) {
+      DBG_OUTPUT_PORT.println("Sent less data than expected!");
+    }
+    file.close();
+    return true;
+  }
+
+  return false;
 }
 
+
+/*
+   As some FS (e.g. LittleFS) delete the parent folder when the last child has been removed,
+   return the path of the closest parent still existing
+*/
+String lastExistingParent(String path) {
+  while (path != "" && !fileSystem->exists(path)) {
+    if (path.lastIndexOf("/") > 0) {
+      path = path.substring(0, path.lastIndexOf("/"));
+    } else {
+      path = String();  // No slash => the top folder does not exist
+    }
+  }
+  DBG_OUTPUT_PORT.println(String("Last existing parent: ") + path);
+  return path;
+}
+
+/*
+   Handle the creation/rename of a new file
+   Operation      | req.responseText
+   ---------------+--------------------------------------------------------------
+   Create file    | parent of created file
+   Create folder  | parent of created folder
+   Rename file    | parent of source file
+   Move file      | parent of source file, or remaining ancestor
+   Rename folder  | parent of source folder
+   Move folder    | parent of source folder, or remaining ancestor
+*/
 void handleFileCreate() {
-  if (server.args() == 0) {
-    return server.send(500, "text/plain", "BAD ARGS");
+  if (!fsOK) {
+    return returnFail("FS INIT ERROR");
   }
-  String path = server.arg(0);
-  DBG_OUTPUT_PORT.println("handleFileCreate: " + path);
+  if (server.args() == 0 || server.args() > 2) {
```

Comment:
Checking `server.args()` will lead to confusion, at least until #6768 which restores the original and expected `::args()` definition is merged.
Indeed in a previous commit from this PR (see comments there), `args()` is increased sometimes.
Can you avoid using it and instead check for the existence (or not) of the expected arguments ?

Reply:
No problem. I rewrote the tests to check for presence of server arguments.